Refactor geometric type converters for improved parsing

AlexRadch opened the pull request:
Close #92424

This commit refactors the parsing logic for string representations of geometric types (Point, Rectangle, Size, SizeF) by using `Span<Range>` to enhance performance and reduce memory allocations.

Exception handling has been improved with more specific error messages for parsing failures.

Additionally, tests have been updated to reflect changes in expected values and formats, covering a wider range of cases, including edge cases and whitespace handling.

The syntax for array initialization has been modernized to use collection initializers for better readability.

Overall, these changes enhance the robustness, readability, and performance of the type conversion code.

AlexRadch committed:
Refactor geometric type converters for improved parsing

This commit refactors the parsing logic for string representations of geometric types (Point, Rectangle, Size, SizeF) by using `Span<Range>` to enhance performance and reduce memory allocations.

Exception handling has been improved with more specific error messages for parsing failures.

Additionally, tests have been updated to reflect changes in expected values and formats, covering a wider range of cases, including edge cases and whitespace handling.

The syntax for array initialization has been modernized to use collection initializers for better readability.

Overall, these changes enhance the robustness, readability, and performance of the type conversion code.

bencyoung-Fignum commented on src/libraries/System.ComponentModel.TypeConverter/src/System/Drawing/SizeConverter.cs:
```diff
@@ -70,16 +67,16 @@ public override bool CanConvertTo(ITypeDescriptorContext? context, [NotNullWhen(
                     // Note: ConvertToString will raise exception if value cannot be converted.
-                    var args = new string?[]
-                    {
+                    string?[] args =
```
`Could this be $"{intConverter.ConvertToString(context, culture, size.Width)}{culture.TextInfo.ListSeparator} {intConverter.ConvertToString(context, culture, size.Height)}"`

to avoid array allocation?

bencyoung-Fignum commented on src/libraries/System.ComponentModel.TypeConverter/src/System/Drawing/SizeFConverter.cs:
```diff
@@ -67,16 +65,16 @@ public override bool CanConvertTo(ITypeDescriptorContext? context, [NotNullWhen(
                     TypeConverter floatConverter = TypeDescriptor.GetConverterTrimUnsafe(typeof(float));
-                    var args = new string?[]
-                    {
+                    string?[] args =
```
Same here

AlexRadch committed:
Refactor CanConvertTo methods for clarity

This commit refactors the `CanConvertTo` method in the `PointConverter`, `RectangleConverter`, `SizeConverter`, and `SizeFConverter` classes.

The changes enhance readability and maintainability by replacing the use of an array to hold converted values with direct string interpolation for constructing the return value. This results in clearer and more concise code.

karakasa commented on src/libraries/System.ComponentModel.TypeConverter/src/System/Drawing/RectangleConverter.cs:
```diff
@@ -66,23 +65,21 @@ public override bool CanConvertTo(ITypeDescriptorContext? context, [NotNullWhen(
-                    ConstructorInfo? ctor = typeof(Rectangle).GetConstructor(new Type[] {
-                        typeof(int), typeof(int), typeof(int), typeof(int)});
+                    ConstructorInfo? ctor = typeof(Rectangle).GetConstructor(
+                        [typeof(int), typeof(int), typeof(int), typeof(int)]);
```
Can these param lists be static readonly arrays?

AlexRadch committed:
Optimize ConvertFrom methods in converter classes

Refactor the `ConvertFrom` method in `PointConverter`, `RectangleConverter`, `SizeConverter`, and `SizeFConverter` to use `AsSpan().Trim()` for improved performance by reducing string allocations.

ericstj commented:
/azp run runtime

ericstj reviewed (approved):
Changes LGTM, thank you for the improvement

ericstj commented:
/ba-g failures are networking test flakiness
